src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java

```diff
@@ -0,0 +1,617 @@
+      RobotMap.REAR_LEFT_TURNING_CAN_ID,
+      DriveCal.BACK_LEFT_CHASSIS_ANGULAR_OFFSET_RAD);
+
+  public final SwerveModule rearRight = new SwerveModule(
+      RobotMap.REAR_RIGHT_DRIVING_CAN_ID,
+      RobotMap.REAR_RIGHT_TURNING_CAN_ID,
+      DriveCal.BACK_RIGHT_CHASSIS_ANGULAR_OFFSET_RAD);
+
+  private final Pigeon2 gyro = new Pigeon2(RobotMap.PIGEON_CAN_ID);
+  private ChassisSpeeds lastSetChassisSpeeds = new ChassisSpeeds(0.0, 0.0, 0.0);
+  public Optional<Pose2d> targetPose = Optional.empty();
+  /** include path planner initialization here */
+  private MedianFilter pitchFilter = new MedianFilter(Constants.PLACEHOLDER_INT);
+  private double latestFilteredPitchDeg = Constants.PLACEHOLDER_DOUBLE;
+
+  SwerveDriveOdometry odometry = new SwerveDriveOdometry(Constants.SwerveDrive.DRIVE_KINEMATICS,
+      Rotation2d.fromDegrees(Constants.PLACEHOLDER_DOUBLE), getModulePositions());
+
+  /** Multiplier for drive speed, does not affect trajectory following */
+  private double throttleMultiplier = 1.0;
+  private BooleanSupplier isTimedMatch;
+
+  public DriveSubsystem(Lights lightsSubsystem, BooleanSupplier isTimedMatchFunc) {
+    /** Need to factory default settings for gyro, but no function exists */
+    gyro.reset();
+    /** Need to configure a mount pose for gyro, but no function exist */
+    this.lights = lightsSubsystem;
+    this.isTimedMatch = isTimedMatchFunc;
+  }
+
+  public double getFilteredPitch() {
+    return latestFilteredPitchDeg - Constants.SwerveSubsystem.IMU_PITCH_BIAS_DEG;
+  }
+
+  @Override
+  public void periodic() {
+    latestFilteredPitchDeg = pitchFilter.calculate(gyro.getPitch().getValue());
+
+    // Update the odometry in the periodic block
+    frontLeft.periodic();
+    frontRight.periodic();
+    rearLeft.periodic();
+    rearRight.periodic();
+    odometry.update(Rotation2d.fromDegrees(gyro.getYaw().getValue()), getModulePositions());
+  }
+
+  public SwerveModulePosition[] getModulePositions() {
+    return new SwerveModulePosition[] {
+        frontLeft.getPosition(),
+        frontRight.getPosition(),
+        rearLeft.getPosition(),
+        rearRight.getPosition()
+    };
+  };
+
+  public void burnFlashSparks() {
+    frontLeft.burnFlashSparks();
+    frontRight.burnFlashSparks();
+    rearLeft.burnFlashSparks();
+    rearRight.burnFlashSparks();
+  }
+
+  /**
+   * Returns the currently-estimated pose of the robot.
+   *
+   * @return The pose.
+   */
+  public Pose2d getPose() {
+    return odometry.getPoseMeters();
+  }
+
+  public ChassisSpeeds getChassisSpeeds() {
+    return lastSetChassisSpeeds;
+  }
+
+  /**
+   * applies the robot-relative output speeds of the FollowPathHolonomic Command
+   * adapted from drive() (but doesn't need the code to turn xSpeed, ySpeed, rot,
+   * fieldRelative into a ChassisSpeeds object)
+   * does not call the correctForDynamics method because we pass in a corrected
+   * lastSetChassisSpeeds (// TODO is this right???)
```

Comment:
I doubt path planner is doing this correction

Reply:
@avakimc